### PR TITLE
[WIP]: Middleware

### PR DESCRIPTION
**Pull request by Wr4thon**

### CODE REVIEW

File: middleware/errorcounter.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			if v, ok := c.Delivery().Headers[errorCounterKey]; ok {
+				if i, ok := v.(int32); ok {
+					c.Set(ErrorCounter{}, i)
+				}
+			} else {
+				c.Set(ErrorCounter{}, int32(0))
+			}
+
+			err := hf(c)
+
```
Comment (lrotermund) [inline middleware/errorcounter.go:31]: unnecessary linebreak

File: middleware/errorcounter_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+func Test_Middleware(t *testing.T) {
+	connection, err := gorabbitmq.NewConnection(gorabbitmq.ConnectionSettings{
+		Host:     "localhost",
+		Password: "guest",
+		UserName: "guest",
+		Port:     5672,
+	}, gorabbitmq.ChannelSettings{
+		UsePrefetch: false,
+	})
+
```
Comment (lrotermund) [inline middleware/errorcounter_test.go]: unnecessary linebreak

File: middleware/errorcounter_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			Durable:          true,
+			Exclusive:        true,
+			NoWait:           false,
+		},
+		gorabbitmq.WithErrorHandler(func(c gorabbitmq.Context, e error) error {
+			return nil
+		}),
+	)
+
+	defer queue.Close()
```
Comment (lrotermund) [inline middleware/errorcounter_test.go]: Possible nil pointer exception. In case of an error queue is nil

File: queue.go [hunk 4]
```diff
@@ -11,71 +11,100 @@ import (
...
+		q.errorHandler = errorHandler
+		return nil
+	}
+}
+
+// WithContextExtractor sets the ContextExtractor function that gets called,
+// before a delivery gets served
+func WithContextExtractor(contextExtractor ContextExtractor) ConfigBuilder {
+	return func(q *queue) error {
+		q.contextExtractor = contextExtractor
```
Comment (lrotermund) [inline queue.go:92]: unnecessary linebreak

File: queue.go [hunk 5]
```diff
@@ -11,71 +11,100 @@ import (
...
 	}
 }
 
-func WithLoggingContextBuilder(contextBuilder LoggingContextBuilder) ConfigBuilder {
+// WithContextBuilder sets the ContextBuilder function that gets called,
+// before a delivery gets sent to the RabbitMQ
+func WithContextBuilder(contextBuilder ContextBuilder) ConfigBuilder {
 	return func(q *queue) error {
-		q.loggingBuilder = contextBuilder
+		q.contextBuilder = contextBuilder
```
Comment (lrotermund) [inline queue.go]: unnecessary linebreak

File: queue.go [hunk 6]
```diff
@@ -149,23 +201,17 @@ func (c *queue) RegisterConsumer(consumerSettings ConsumerSettings, deliveryCons
...
-		if err != nil {
-			return errors.Wrap(err, "error while loading context")
-		}
-
-		err = deliveryConsumer(ctx, item)
-
-		if err != nil {
-			return errors.Wrap(err, "error while delivering message to consumer")
+		err := c.consumeItem(item, deliveryConsumer, middleware)
+		if err != nil && c.errorChan != nil {
```
Comment (lrotermund) [inline queue.go:203]: What happens in the else case? Does the error disappear, or should an error be returned?

File: queue.go [hunk 7]
```diff
@@ -213,17 +260,35 @@ func (c *queue) ConsumerOnce(consumerSettings ConsumerSettings, deliveryConsumer
...
+	queueContext, err := c.loadContext(item)
 	if err != nil {
 		return errors.Wrap(err, "error while loading context")
 	}
 
-	err = deliveryConsumer(ctx, item)
+	h := applyMiddleware(deliveryConsumer, middleware...)
+
+	err = h(queueContext)
+
```
Comment (lrotermund) [inline queue.go]: unnecessary linebreak

File: queue_test.go [hunk 8]
```diff
@@ -205,3 +204,83 @@ func TestSendWithContext(t *testing.T) {
...
+func Test_Middleware(t *testing.T) {
+	connection, err := NewConnection(ConnectionSettings{
+		Host:     "localhost",
+		Password: "guest",
+		UserName: "guest",
+		Port:     5672,
+	}, ChannelSettings{
+		UsePrefetch: false,
+	})
+
```
Comment (lrotermund) [inline queue_test.go]: unnecessary linebreak

File: queue_test.go [hunk 9]
```diff
@@ -205,3 +204,83 @@ func TestSendWithContext(t *testing.T) {
...
+	}
+
+	queue, err := connection.ConnectToQueue(QueueSettings{
+		QueueName:        "test",
+		DeleteWhenUnused: true,
+		Durable:          false,
+		Exclusive:        true,
+		NoWait:           false,
+	})
+
```
Comment (lrotermund) [inline queue_test.go]: unnecessary linebreak

File: middleware/errorcounter_test.go (hunk 3 above)
Comment (Wr4thon) [reply middleware/errorcounter_test.go]: well, it's a test, so if it fails, you have to look anyway.
if there is nil value, obviously the logic changed.

File: queue.go (hunk 6 above)
Comment (Wr4thon) [reply queue.go:203]: if there is an error, and you want to get notified, just use the error channel.

**Pull request merged**